File: 287-find-the-duplicate-number/287-find-the-duplicate-number.scala
object Solution {
    def findDuplicate(nums: Array[Int]): Int = {
        var fast = nums(0)
        var slow = nums(0)

        var firstStep = true

        while(firstStep || fast != slow) {
            firstStep = false
            fast = nums(nums(fast))
            slow = nums(slow)
        }

        fast = nums(0)

        while(fast != slow) {
            slow = nums(slow)
            fast = nums(fast)
        }

        fast 
    }
}

File: 287-find-the-duplicate-number/287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int fast = nums[nums[nums[0]]];
        int slow = nums[nums[0]];
        
        while(fast != slow) {
            fast = nums[nums[fast]];
            slow = nums[slow];
        }
        
        fast = nums[0];
        
        while(fast != slow) {
            fast = nums[fast];
            slow = nums[slow];
        }
        
        return slow;
    }
}